### PR TITLE
[Relationships] Hasmany morphmany

### PR DESCRIPTION
**Pull request by pxpm**
## WHY

Only `HasMany` was working and only allow some specific scenario. (select previously created entries). We should make `HasMany/MorphMany` consistent as they are the same relation, the difference is the same as `HasOne/MorphOne`, for what matters they are the same relation at the end.

### BEFORE - What was wrong? What was happening before this PR?

- You cannot create HasMany/MorphMany entries. 
- You cannot "not delete" the related entry when unselecting.

### AFTER - What is happening after this PR?

You can do all the above.

### How did you achieve that, in technical terms?

Implemented `creatable` and `selectable` concepts into HasMany/MorphMany relations. So you can still use a select for a list or create the many relation "on the fly". 

Added `fallback_id` and `force_delete` keys into selectable relations. 

By default `force_delete` is `true`. When unselecting the entries from a select they would be deleted from the related table.
Setting `force_delete` to `false` would make Backpack set `null` in the relation in case the related column is nullable. 
`fallback_id` takes precedence over the others, if defined Backpack will set the connecting key to whatever developer provided. 

Added tests for all that I mentioned.

### Is it a breaking change or non-breaking change?

Probably non-breaking, but better be safe go 4.2.

### How can we test the before & after?

Difficult because this concepts didn't exist in previous versions. But the new concept is available in 4.2 branch of demo.

PS: @tabacitu the if/elses could probably use some refactor, I think this way they are more explatatory of what is happening in the app lifecycle and easier to follow atleast for me :)

### CODE REVIEW

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 1]
```diff
@@ -169,14 +171,115 @@ private function createRelationsForItem($item, $formattedRelations)
...
+                if (! $relation_column_is_nullable || $force_delete) {
+                    $model_instance->whereNotIn($model_instance->getKeyName(), $relation_values)
+                        ->where($relation_foreign_key, $item->{$relation_local_key})
+                        ->delete();
+                } else {
+                    $model_instance->whereNotIn($model_instance->getKeyName(), $relation_values)
+                        ->where($relation_foreign_key, $item->{$relation_local_key})
+                        ->update([$relation_foreign_key => null]);
+                }
+            }
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Create.php:233]:
```suggestion
            $removed_entries = $model_instance->whereNotIn($model_instance->getKeyName(), $relation_values)->where($relation_foreign_key, $item->{$relation_local_key});
            if ($relationDetails['fallback_id'] !== false) {
                    $removed_entries->update([$relation_foreign_key => $relationDetails['fallback_id']]);
            } else {
                if (! $relation_column_is_nullable || $force_delete) {
                    $removed_entries->delete();
                } else {
                    $removed_entries->update([$relation_foreign_key => null]);
                }
            }
```

Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:233]: I see the same query is made 3 times here. We can move it before the conditional to only do it once. Not sure I understood correctly what it does, and chose a good variable name.

File: src/resources/views/crud/fields/relationship/relationship_select.blade.php [hunk 2]
```diff
@@ -50,7 +50,7 @@
 
 @include('crud::fields.inc.wrapper_start')
     <label>{!! $field['label'] !!}</label>
-
+    <input type="hidden" name="{{ $field['name'] }}" value="" @if(in_array('disabled', $field['attributes'] ?? [])) disabled @endif />
```
Comment (tabacitu) [inline src/resources/views/crud/fields/relationship/relationship_select.blade.php:53]: @pxpm could you please explain what this does? I see it's been added in 3 fields but can't tell what it does and why it exists.

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 3]
```diff
@@ -218,6 +321,8 @@ private function getRelationDetailsFromInput($input)
             // we make sure that at least once we set the relation details
             $fieldDetails = Arr::get($relationDetails, 'relations.'.$key, []);
             $fieldDetails['model'] = $fieldDetails['model'] ?? $field['model'];
+            $fieldDetails['fallback_id'] = $fieldDetails['fallback_id'] ?? $field['fallback_id'] ?? false;
+            $fieldDetails['force_delete'] = $fieldDetails['force_delete'] ?? $field['force_delete'] ?? true;
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Create.php:325]: Ugh... do we really have to add these attributes for ALL relationships, when they're only used by two of them? Then again, this is one place, if we add them to the fields that use them... it'd be 3 places right?... 😔

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 4]
```diff
@@ -169,14 +171,115 @@ private function createRelationsForItem($item, $formattedRelations)
...
         }
     }
 
+    /**
+     * When using the HasMany/MorphMany relations as selectable elements we use this function to "mimic-sync" in those relations.
+     * Since HasMany/MorphMany does not have the `sync` method, we manually re-create it.
+     * Here we add the entries that developer added and remove the ones that are not in the list.
+     * This removal process happens with the following rules:
+     * - by default Backpack will behave like a `sync` from M-M relations: it deletes previous entries and add only the current ones.
+     * - `force_delete` is configurable in the field, it's `true` by default. When false, if connecting column is nullable instead of deleting the row we set the column to null.
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Create.php:196]: Are we _sure_ we want to `force_delete` by default? I imagine this scenario. I have a tag with A BUNCH of articles. One Tag HasMany Articles right? 

I go to the tag EDIT operation. This PR will make it so I have a new input there, which will allow me to select the Articles for this Tag, correct. Do I understand correctly that when I remove an Article association from a Tag... that Article will be deleted?

If so... I don't think that's the expected behavior...

Or have I greatly misunderstood this new functionality? 👀

File: src/resources/views/crud/fields/relationship/fetch_or_create.blade.php [hunk 5]
```diff
@@ -101,50 +101,51 @@
...
-            @foreach ($current_value as $key => $item)
-                <option value="{{ $key }}" selected>
-                    {{ $item }}
-                </option>
-            @endforeach
-        @endif
-
-    </select>
+        <input type="hidden" name="{{ $field['name'] }}" value="" @if(in_array('disabled', $field['attributes'] ?? [])) disabled @endif />    
+        <select
```
Comment (tabacitu) [inline src/resources/views/crud/fields/relationship/fetch_or_create.blade.php:105]: Am I correc that nothing in this SELECT changed? It's just the input above that's new? I can't for the life of me spot any differences, it looks like a diff mistake right?

**Review by tabacitu — CHANGES_REQUESTED** [review]
Thank you Pedro - this one is much _much_ cleaner, don't have much to say in terms of code style. But I do have a few questions in terms of how it works.

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 4 above)
Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Create.php:196]: Hey @tabacitu , the relation setup is the following:

HasMany <-> BelongsTo

So, you have an Article -> HasMany -> Tags | Tag -> BelongsTo -> Article (via article_id in tags table).

If you are editing an Article you have a select for tags. When those tags get "unselected" they get deleted. If `force_delete = false` and column allows nullable we will set `article_id` to null in the tags table. Same for `fallback_id` you get the idea. If your tags can belong to more than one article you are using the incorrect relation should be `BelongsToMany` or `MorphToMany`

The reasoning to make it false by default, is that changing the value to "null" it's a developer feature (what is expected is that you have a table with "related keys" set so `article_id` really references an article) and people expecting this but didn't make the column non-nullable or didn't setup foreignKeys would end-up with "dead entries" in the database probably without noticing it.
Developers aware of it and that want to deal with those entries can setup the `force_delete => false` (if the field is nullable) otherwise `fallback_id => 0` for example for non-nullable columns.

I think it makes more sense, but you know... I migth be wrong!

Let me know if I expressed myself in an understandable manner! 👍

File: src/resources/views/crud/fields/relationship/fetch_or_create.blade.php (hunk 5 above)
Comment (pxpm) [reply src/resources/views/crud/fields/relationship/fetch_or_create.blade.php:105]: Indeed, only the hidden input were added and also indentation for the `<select>` tag, I guess that is what's detected as a change, the identation. 

About the hidden input it should not be here in this PR, but I was about to submit it in a separate one so you can test it out. 

But it's easy to reproduce:
- go to `relationship_select` and remove the hidden input. Add a multiple relationship WITHOUT ajax to use that field. 
- Select one or more and save.
- go edit the entry and clear the select input by using the `x` in each individual option, not the `clear` button.
- save and you will see nothing happens (because when cleared it would not post NOTHING so will not be present in the request.) - IT WORKS with the clear button tho. 

I will be removing this from here and create a separate PR, maybe we can fix it in JS on the field side and don't need to add the hidden inputs. Thing is, with hidden input it works.

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 1 above)
Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Create.php]: Done! 👍

File: src/resources/views/crud/fields/relationship/relationship_select.blade.php (hunk 2 above)
Comment (pxpm) [reply src/resources/views/crud/fields/relationship/relationship_select.blade.php]: I explained in comment: https://github.com/Laravel-Backpack/CRUD/pull/4016#discussion_r771464955

**Comment by pxpm** [comment]
@tabacitu refactored the codestyle and removed the hidden inputs. 

Please see: https://github.com/Laravel-Backpack/CRUD/pull/4016#discussion_r771464955 for more detail about it.

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 4 above)
Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:196]: Wait wait wait. My example doesn't make sense, nor does yours 😅  Let's use realistic examples, I think that's the source of confusion there:
- over here we're talking about `HasMany` and `MorphMany` (1-n);
- my example showed Articles & Tags, which is a bad example, because realistically between Article and Tag the relationship would be `belongsToMany` (n-n)
- your example with [Article `hasMany` Tags] is a bad example too, it's difficult to follow because never in a db will [Article `belongsTo` Tag]; 

So I'm going to try to rewrite your answer above from [Article `hasMany` Tags] to [Author  `hasMany` Books] and [Book `belongsTo` Author] (Article becomes Author, Tag becomes Book). Hopefully that way I can understand 😅 

>So, you have an Author -> HasMany -> Books | Book -> BelongsTo -> Author (via author_id in `books` table).

> If you are editing an Author you have a select for books. When those books get "unselected" they get deleted. If `force_delete = false` and column allows nullable we will set `author_id` to null in the books table. Same for `fallback_id` you get the idea. If your books can belong to more than one author you are using the incorrect relation should be `BelongsToMany` or `MorphToMany`

> The reasoning to make it false by default, is that changing the value to "null" it's a developer feature (what is expected is that you have a table with "related keys" set so `author_id` really references an article) and people expecting this but didn't make the column non-nullable or didn't setup foreignKeys would end-up with "dead entries" in the database probably without noticing it.
> Developers aware of it and that want to deal with those entries can setup the `force_delete => false` (if the field is nullable) otherwise `fallback_id => 0` for example for non-nullable columns.

> I think it makes more sense, but you know... I might be wrong!

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 6]
```diff
@@ -169,14 +171,111 @@ private function createRelationsForItem($item, $formattedRelations)
...
+     * - `fallback_id` could be provided. In this case instead of deleting we set the connecting key to whatever developer gives us.
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Create.php:197]: How will `fallback_id` play along with Laravel's `withDefault()` and the DB column `default`?

```php
// let's remember the dev might have done this to have
// an author called "Unknown Author" with id 1022
public function author()
{
    return $this->belongsTo('App\Author')->withDefault(1022);
}

// or inside their migration they might have set a `default` on the column
$table->integer('author_id')->default(1022);
```

From what I understand, the `fallback_id` will make Backpack set the value to that, so it won't even reach `withDefault()` or the `default` from the DB, is that correct? If so, that's great, I think that's the expected behavior.

Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:197]: But if `fallback_id` is NOT specified, I think what should happen is this, in this order:
- if column is `NOT NULL`, delete 🛑
- if column is `nullable`
    - if a default has been set on the `belongsTo` relationship, set it to that;
    - if a default has been set on the db column, set it to that;
    - if no default on the db column... it's the same as setting the default to `null`, so set it to that;

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 4 above)
Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:196]: OK so I think I understand now, and from rephrasing the example above we're on the same page. I see a few tricky issues here, and we should be very _very_ careful about them. I'll add them in different places in this review, though, so that we can talk about them individually.

Here, I'll continue with my opinion for the `force_delete` default: **by default, it should be as non-destructive as possible**. Yes, the developer might have made a mistake when setting his column to `nullable`. But it's _safer_ to assume they did know what they were doing:
- [ ] if column is `nullable`, it MUST become whatever the `default` is on that relationship or db column... if there is no default, then it will become `null`; the user can easily override this default behaviour, using `force_delete => true` or `fallback_id => 141`; let's default to non-destructive! we can deal with orphaned entries in the docs with a big fat notice, since it's a problem... but not a big one; the dev can easily delete all orphaned entries in one SQL line, 10 years after they created the admin panel - no harm was done, just bloat which is easily identifiable;
- [ ] if column is `NOT NULL`, it should _probably_ be deleted by default; but we're still not 100% sure about that; both developer and admin should be cautioned that this will delete something - in a way they cannot miss it; THEN we can be sure we can delete the related entry; but how can we caution them?
    - [ ] for the admin
         - (A) every time the X is clicked, we could show an alert saying "_Are you sure? After you save this form, that related entry will be deleted._"; but is there a _simple_ way to add that alert, that will not bloat up our existing fields? or...
         - (B) any time `force_delete => true` we add a hint to that field (maybe yellow with `text-warning`), saying "_ATTENTION: Any entries you remove here will be deleted after you submit this form._"; this would probably be easier to implement, and make it easy for the dev to override (just change the hint); but warning hints only cover maybe 70% of all people... I know for a fact... I see it everyday in the Backpack "free license" form; people just don't read it, even if it's red;
    - [ ] for the developer, there is only one way to be certain they know the related entry will be deleted... and that's to _only do it if they tell us_; if we set `force_delete` to `false` by default, and only delete if the dev explicitly told us to delete, my making it `true`; if you haven't set that to true... when you test it... it will probably try to set null on a non-nullable column and error; this has one downside... the column won't "_just work_", you'll have to define `force_delete => true` for it to work in this particular scenario;

What do you think in terms of UI and DI @pxpm ? Personally I think:
- [ ] the yellow hint MUST be added if `force_delete` is true; it will let both the dev and most admins know what will happen;
- [ ] the alert probably SHOULD be added, to make absolutely sure admins don't delete stuff by mistake;
- [ ] the `force_delete` COULD be `false` by default; though... if we add the yellow hint... the dev will definitely notice it when they test it so... 🤷‍♂️ maybe this one is actually a WON'T;

What do you think?

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 3 above)
Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:321]: Right right right!!! If this is only ever going to be used inside the `relationship` field... where we already have a switchboard depending on the relation type... let's set `force_delete` and `fallback_id` there, inside the switchboard.

That way, the rest of the relationships fields won't have these `force_delete` and `fallback_id` attributes, which would do NOTHING for them anyway. No more bloat for them 🎉

Right @pxpm ?

File: src/resources/views/crud/fields/relationship/fetch_or_create.blade.php (hunk 5 above)
Comment (tabacitu) [reply src/resources/views/crud/fields/relationship/fetch_or_create.blade.php:104]: Oh I get it, thanks. Yes, separate PR would be better, please 🙏

Comment (tabacitu) [reply src/resources/views/crud/fields/relationship/fetch_or_create.blade.php:104]: I see you've already done that so feel free to click Resolve on this.

**Comment by tabacitu** [comment]
Thank you @pxpm ! I think I understand much better now. I've replied inline, let's continue the conversation there, as there are a few tricky things we need to worry about - but I think we have solutions for all of them.

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 4 above)
Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Create.php:196]: @tabacitu I know we probably gona discuss this today, but since I read your reply now and don't want to forget what I thougth in a few hours I will leave it written here:
- I am quite a bit convinced that you are right in terms of making `force_delete => false` by default.
- I totally don't agree with notices or alerts on panel. Reasoning:
     - This is a developer configuration, not a "user should be aware of" feature. 1st developer creates the panel with their rules and only after user is going to use the panel. If developer set it up to be deleted there is no need to "warn" the user that the entry is going to be deleted, that's how developer configured the panel and you know `Devs know better!` 🙃

Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:196]: Yeah... ok... I guess if we do `'force_delete' => false` by default, in order for the dev to make this delete stuff they have to read the comment or docs. And in there we can make it super-clear to them what using it will entail.

Deal.

**Comment by pxpm** [comment]
@tabacitu refactored and added some more tests . Anything you see is missing to cover ? 

Let me know I still feel we could probably rename something here and there but I am vicious rn!

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 7]
```diff
@@ -169,14 +171,114 @@ private function createRelationsForItem($item, $formattedRelations)
...
+     *
+     * @return void
+     */
+    public function attachManyRelation($item, $relation, $relationDetails, $relation_values)
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Create.php:201]:
```suggestion
    private function attachManyRelation($item, $relation, $relationDetails, $relation_values)
```

From what I see, this method is only called inside `createRelationsForItem()`, which itself is on the CrudPanel object and `private`...

Any benefit to having this `public` @pxpm ?

Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:201]: I'm pretty sure we're going to refractor this in the next 12 months, so it'd be better to only add `private` methods, so we can do that without a breaking change.

File: src/app/Models/Traits/HasRelationshipFields.php [hunk 8]
```diff
@@ -64,11 +64,7 @@ public function getColumnType($columnName)
      */
     public static function isColumnNullable($column_name)
     {
-        // create an instance of the model to be able to get the table name
-        $instance = new static();
-
-        $conn = $instance->getConnectionWithExtraTypeMappings();
-        $table = $instance->getTableWithPrefix();
+        [$conn, $table] = self::getModelConnectionAndTable();
```
Comment (tabacitu) [inline src/app/Models/Traits/HasRelationshipFields.php:67]:
```suggestion
        [$conn, $table] = self::getConnectionAndTable();
```

File: src/app/Models/Traits/HasRelationshipFields.php [hunk 9]
```diff
@@ -86,4 +82,54 @@ public static function isColumnNullable($column_name)
...
+    {
+        [$conn, $table] = self::getModelConnectionAndTable();
+
+        return $conn->getDoctrineColumn($table, $column_name)->getDefault();
+    }
+
+    /**
+     * Return the current model connection and table name.
+     */
+    private static function getModelConnectionAndTable()
```
Comment (tabacitu) [inline src/app/Models/Traits/HasRelationshipFields.php:127]:
```suggestion
    private static function getConnectionAndTable()
```

File: src/app/Models/Traits/HasRelationshipFields.php [hunk 10]
```diff
@@ -86,4 +82,54 @@ public static function isColumnNullable($column_name)
...
+
+    /**
+     * Return the db column default value.
+     *
+     * @param  string  $column_name  The name of the db column.
+     * @return bool
+     */
+    public static function getDbColumnDefaultValue($column_name)
+    {
+        [$conn, $table] = self::getModelConnectionAndTable();
```
Comment (tabacitu) [inline src/app/Models/Traits/HasRelationshipFields.php:119]:
```suggestion
        [$conn, $table] = self::getConnectionAndTable();
```

File: src/app/Models/Traits/HasRelationshipFields.php [hunk 11]
```diff
@@ -86,4 +82,54 @@ public static function isColumnNullable($column_name)
...
+
+    /**
+     * Checks if the given column name has default value set.
+     *
+     * @param  string  $column_name  The name of the db column.
+     * @return bool
+     */
+    public static function dbColumnHasDefaultValue($column_name)
+    {
+        [$conn, $table] = self::getModelConnectionAndTable();
```
Comment (tabacitu) [inline src/app/Models/Traits/HasRelationshipFields.php:94]:
```suggestion
        [$conn, $table] = self::getConnectionAndTable();
```

File: src/app/Models/Traits/HasRelationshipFields.php [hunk 12]
```diff
@@ -86,4 +82,54 @@ public static function isColumnNullable($column_name)
...
         }
     }
+
+    /**
+     * Checks if the given column name has default value set.
+     *
+     * @param  string  $column_name  The name of the db column.
+     * @return bool
+     */
+    public static function dbColumnHasDefaultValue($column_name)
```
Comment (tabacitu) [inline src/app/Models/Traits/HasRelationshipFields.php:92]:
```suggestion
    public static function dbColumnHasDefault($column_name)
```

File: src/app/Models/Traits/HasRelationshipFields.php [hunk 13]
```diff
@@ -86,4 +82,54 @@ public static function isColumnNullable($column_name)
...
+        }
+    }
+
+    /**
+     * Return the db column default value.
+     *
+     * @param  string  $column_name  The name of the db column.
+     * @return bool
+     */
+    public static function getDbColumnDefaultValue($column_name)
```
Comment (tabacitu) [inline src/app/Models/Traits/HasRelationshipFields.php:117]:
```suggestion
    public static function getDbColumnDefault($column_name)
```

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 14]
```diff
@@ -169,14 +171,114 @@ private function createRelationsForItem($item, $formattedRelations)
...
+
+        if (isset($relationDetails['fallback_id']) && $relationDetails['fallback_id'] !== false) {
+            $removed_entries->update([$relation_foreign_key => $relationDetails['fallback_id']]);
+        } else {
+            if ($force_delete) {
+                $removed_entries->delete();
+            // if column is not nullable we will check if there is some column default
+            // provided in database schema. If there is, we use it as fallback.
+            } elseif (! $relation_column_is_nullable) {
+                if ($model_instance->dbColumnHasDefaultValue($relation_foreign_key)) {
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Create.php:241]:
```suggestion
                if ($model_instance->dbColumnHasDefault($relation_foreign_key)) {
```

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 15]
```diff
@@ -169,14 +171,114 @@ private function createRelationsForItem($item, $formattedRelations)
...
+                    $removed_entries->update([$relation_foreign_key => $model_instance->getDbColumnDefaultValue($relation_foreign_key)]);
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Create.php:242]:
```suggestion
                    $removed_entries->update([$relation_foreign_key => $model_instance->getDbColumnDefault($relation_foreign_key)]);
```

**Review by tabacitu — CHANGES_REQUESTED** [review]
Did some renaming, optimizing for simplicity, going to merge them here.

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 16]
```diff
@@ -169,14 +171,114 @@ private function createRelationsForItem($item, $formattedRelations)
...
+    }
+
+    /**
+     * Handle HasMany/MorphMany relations when used as creatable entries in the crud.
+     * By using repeatable field, developer can allow the creation of such entries
+     * in the crud forms.
+     *
+     * @return void
+     */
+    public function createManyEntries($entry, $relation, $relationMethod, $relationDetails)
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Create.php:257]:
```suggestion
    private function createManyEntries($entry, $relation, $relationMethod, $relationDetails)
```

**Review by tabacitu — CHANGES_REQUESTED** [review]

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 17]
```diff
@@ -169,14 +171,113 @@ private function createRelationsForItem($item, $formattedRelations)
...
+     * When using the HasMany/MorphMany relations as selectable elements we use this function to "mimic-sync" in those relations.
+     * Since HasMany/MorphMany does not have the `sync` method, we manually re-create it.
+     * Here we add the entries that developer added and remove the ones that are not in the list.
+     * This removal process happens with the following rules:
+     * - by default Backpack will behave like a `sync` from M-M relations: it deletes previous entries and add only the current ones.
+     * - `force_delete` is configurable in the field, it's `true` by default. When false, if connecting column is nullable instead of deleting the row we set the column to null.
+     * - `fallback_id` could be provided. In this case instead of deleting we set the connecting key to whatever developer gives us.
+     *
+     * @return void
+     */
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Create.php:200]: Found a bug: **clearing the select does nothing**. I suspect this is in connection with another bug, which will be fixed separately, but wanted to confirm. @pxpm should we fix this now?

Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:200]: Note that if I only remove one item, it will be saved ok (falls back to `fallback_id` or deletes if `force_delete`).

File: src/app/Library/CrudPanel/Traits/Create.php [hunk 18]
```diff
@@ -169,14 +171,113 @@ private function createRelationsForItem($item, $formattedRelations)
...
+            return $removed_entries->update([$relation_foreign_key => $model_instance->getDbColumnDefault($relation_foreign_key)]);
+        }
+
+        return $removed_entries->update([$relation_foreign_key => null]);
+    }
+
+    /**
+     * Handle HasMany/MorphMany relations when used as creatable entries in the crud.
+     * By using repeatable field, developer can allow the creation of such entries
+     * in the crud forms.
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Create.php:252]: I've tested this and... as it is right now... I think it poses a serious problem to data integrity. Every time the "monster" form is saved, the IDs on the related HasMany entries will be changed. No matter if there have been changes to the actual `HasMany` repeatable input or not.



This is not ok. From what I understand, this might be done in order to reorder them... Which is a tricky thing to achieve without an `order` column. But this current implementation is not ok. IDs should never be changed. And especially not from a related entity.

I know we said this is _probably_ going to be used for subitems that are completely dependant on the main item, but that doesn't mean OTHER items could not depend on this subitem.

Here's an example:
- Invoice HasMany InvoiceItems
- License BelongsTo InvoiceItem (in order to get the purchase price, purchaser, etc) 

So the `licenses` table has a column `invoice_item_id`. If that ID gets changed on the `invoice_items` table, when you edit your `invoice`... You get an orphaned License.

Any reason to do this @pxpm ?

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 17 above)
Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:200]: Let's make sure to add a unit test for this too, please 🙏 It's an easy use case to forget when we manually test.

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 18 above)
Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:252]: What's particularly odd is that this DOES happen for `HasMany` in the 4.2 demo, but does NOT happen for the `MorphMany` in the demo. Even though I believed they both ran the same code to do this. Notice I will reorder both:
- MorphMany reordering will not be saved
- HasMany reordering will be saved (because IDs get changed)

Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:252]: Again, I think it is WAAAAY more important here for the IDs not to get changed. We can disable the reorder functionality in this use case, for what I care. But the IDs should stay the same.

**Review by tabacitu — CHANGES_REQUESTED** [review]
@pxpm I found some big booboos. Mind taking a look at the comments and telling me what you think, how we should proceed?

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 17 above)
Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Create.php:200]: Hey @tabacitu !

I've added a test for this. In "core" it works as expected. Problem is with select2. (what gets sent when clearing all entries in the "x") I've reported that to you in other comment, and that's why I previously added the hidden inputs. I will have a look if we can fix this in js select2 events.

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 18 above)
Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Create.php:252]: Totally agree with you. 

Although it added the "ability" to order, was not an intentional behavior and a nasty bug, I 100% agree with you that keys should not change. 

I think I fixed that behaviour, and added tests to account for the key too, I was only testing the value of other fields. 

Give it a go please and let me know if it's fixed. Ofcourse, no order for now.

Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:252]: Yup, works fine now! I've pushed a change to `repeatable` on this branch, so we can turn off the reordering. And turned off the reordering inside `repeatable_relation`. So the field is no longer confusing now - it does what it seems to do. And it's no longer dangerous now - it does not increment the IDs 🎉🎉🎉

Good work team 😅

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 17 above)
Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:200]: I have a feeling the JS solution will be 20+ lines long, so way more complicated than the hidden input.

Why is this happening for `HasMany` & `MorphMany`... and not for other relationships? I see `BelongsToMany` and `MorphToMany` are also multiple and are using the `relationship_select` field.

Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Create.php:200]: @tabacitu exactly about this that I wanted to talk to you. 

So here is the problem: 
- BelongsToMany/MorphMany assumes that when not present in the request they should be deleted. You can't use disabled fields for this relationships otherwise when submitting the form it would clear the relationship. 

- HasMany/MorphMany are created as part of other function where we excude fields that are not present in the form input.
    - So like we know, multiple fields does not post input when empty. 
    - I found a solution that is:
 ```php
.on('select2:unselect', function(e) {
                        if ($multiple && $(this).val().length == 0) {
                            $(this).val(null).trigger('change');
                        }});

// this will send on request:
'entity' => [0 => null]
```
We can check if `entity[0] === null` to know that the field was empty, because when it has value `entity[0]` will never be null unless input is empty.

The reason why it worked with clear and not with the "x" was because "clear" automatically calls the JS value change while we need to manually call when removing in the "x". 

BUT, and I think this is a big BUT: we should not add the empty option `-` on select multiples, and if you remove the empty option from the multiple selects, neither clear or "x" can empty it, because it behaves like a natural select multiple, that will not post the value when empty. 

It should be:
```php
@if ($field['allows_null'] && !$field['multiple'])
            <option value="">-</option>
        @endif
```

To sum up, I think the only reasonable and working solution for this is adding the hidden inputs. that will send a "null" value when the input is cleared, also don't post nothing if input is disabled. 

Let me know if I answered your questions,
Pedro

Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Create.php:200]: @tabacitu I've created https://github.com/Laravel-Backpack/CRUD/pull/4025

Has this PR merged inside, and the fix (hidden input). Also is pointing to merge here if you agree.

Only did `relationship_select` so you can test it with the current setup in 4.2 demo branch. 

Let me know how I should proceed!

Pedro

Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:200]: > BelongsToMany/MorphMany assumes that when not present in the request they should be deleted. You can't use disabled fields for this relationships otherwise when submitting the form it would clear the relationship.

(1) So basically... `BelongsToMany` & `MorphToMany` had a bug about this too? If you set it `disabled` instead of `readonly` it would clear all relations on Save? And the hidden input will solve the problems in `BelongsToMany` & `MorphToMany`, but also in `HasMany` and `MorphMany`? We hit 4 birds with one stone?

> To sum up, I think the only reasonable and working solution for this is adding the hidden inputs. that will send a "null" value when the input is cleared, also don't post nothing if input is disabled.

Yup, I think I agree. Hidden input sounds like the best solution to me too... I'll take a look at the PR right away, but please answer my question above too.

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 18 above)
Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Create.php:252]: It's sad that we don't have the reorder functionality here too... 😔 I've added my idea on how to implement this to https://github.com/Laravel-Backpack/CRUD/issues/4012

**Comment by tabacitu** [comment]
Merged 🎉  Thank you for your patience @pxpm . Please see my inline comments then click "Resolve".

What's next? 😀

File: src/app/Library/CrudPanel/Traits/Create.php (hunk 17 above)
Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Create.php:200]: @tabacitu we should apply this hidden inputs to ALL selects that are multiple or can be multiple. And it will fix for everything and the behaviour will be the same for everything. 

So technically you kill all birds. I will submit a PR for your review.

**Comment by tabacitu** [comment]
Documented in https://github.com/Laravel-Backpack/docs/pull/284
